File: leetcode/000307/solution.cpp
class NumArray {
public:
    NumArray(vector<int>& nums) {
        int n = nums.size();
        seg_tree.resize(2 * n);
        for (int i = 0; i < n; i++) {
            seg_tree[n + i] = nums[i];
        }
        for (int i = n - 1; i > 0; i--) {
            seg_tree[i] = seg_tree[2 * i] + seg_tree[2 * i + 1];
        }
    }
    
    void update(int index, int val) {
        int n = seg_tree.size() / 2;
        int i = n + index;
        seg_tree[i] = val;
        while (i > 1) {
            seg_tree[i / 2] = seg_tree[i] + seg_tree[i ^ 1];
            i = i / 2;
        }
    }
    
    int sumRange(int left, int right) {
        int n = seg_tree.size() / 2;
        int l = n + left;
        int r = n + right;
        int ans = 0;
        while (l <= r) {
            if (l & 1) {
                ans += seg_tree[l];
                l = (l + 1) / 2;
            } else {
                l = l / 2;
            }
            if ((r & 1) == 0) {
                ans += seg_tree[r];
                r = (r - 1) / 2;
            } else {
                r = r / 2;
            }
        }
        return ans;
    }
private:
    vector<int> seg_tree;
};

/**
 * Your NumArray object will be instantiated and called as such:
 * NumArray* obj = new NumArray(nums);
 * obj->update(index,val);
 * int param_2 = obj->sumRange(left,right);
 */

File: leetcode/000307/solution.py
class NumArray:

    def __init__(self, nums: List[int]):
        self.n = len(nums)
        self.seg_tree = [0 for _ in range(2*self.n)]
        for i in range(self.n):
            self.seg_tree[self.n+i] = nums[i]
        for i in range(self.n-1, -1, -1):
            self.seg_tree[i] = self.seg_tree[2*i] + self.seg_tree[2*i+1]

    def update(self, index: int, val: int) -> None:
        i = self.n+index
        self.seg_tree[i] = val
        while i > 0:
            l = i
            r = i+1
            
            # if `i` is the right child of its parent
            if i&1 == 1:
                l = i-1
                r = i
                
            self.seg_tree[i//2] = self.seg_tree[l] + self.seg_tree[r]
            i //= 2

    def sumRange(self, left: int, right: int) -> int:
        l = self.n+left
        r = self.n+right
        ans = 0
        while l <= r:
            if l&1 == 1:
                ans += self.seg_tree[l]
                l += 1
            l //= 2
            if r&1 == 0:
                ans += self.seg_tree[r]
                r -= 1
            r //= 2
        return ans


# Your NumArray object will be instantiated and called as such:
# obj = NumArray(nums)
# obj.update(index,val)
# param_2 = obj.sumRange(left,right)
